### PR TITLE
[Refactor] Refactor SemiBaseDetector and SoftTeacher

### PR DESCRIPTION
**Pull request by Czm369**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet/models/detectors/semi_base.py [hunk 1]
```diff
@@ -69,6 +69,21 @@ def reweight_loss(losses: dict, weight: float) -> dict:
                     losses[name] = loss * weight
         return losses
 
+    @staticmethod
```
Comment (ZwwWayne) [inline mmdet/models/detectors/semi_base.py:72]: we can implement a util function in utils and use it in places outside semi-detector

File: mmdet/models/detectors/semi_base.py [hunk 2]
```diff
@@ -69,6 +69,21 @@ def reweight_loss(losses: dict, weight: float) -> dict:
...
+    def rename_loss(prefix: str, losses: dict) -> dict:
+        """Rename loss for different branches."""
+        return {prefix + k: v for k, v in losses.items()}
+
+    @staticmethod
+    def filter_pseudo_instances_by_scores(batch_data_samples: SampleList,
```
Comment (ZwwWayne) [inline mmdet/models/detectors/semi_base.py:78]: filter_gt_instances_by_score

File: mmdet/models/detectors/soft_teacher.py [hunk 3]
```diff
@@ -69,20 +69,17 @@ def loss_by_pseudo_instances(self,
 
         x = self.student.extract_feat(batch_inputs)
 
-        losses = {}
+        pseudo_losses = {}
```
Comment (ZwwWayne) [inline mmdet/models/detectors/soft_teacher.py:72]: we do not need to rename it as pseudo_losses, we can simply use losses and return the renamed ones

File: mmdet/models/detectors/semi_base.py (hunk 1 above)
Comment (Czm369) [reply mmdet/models/detectors/semi_base.py:72]: There exits a similar function in https://github.com/open-mmlab/mmdetection/blob/c9b47820d637ab341ca3ca869116fc3d91030c74/mmdet/models/utils/misc.py#L292, this function is just convenient called by subclasses of semi-detector.

File: mmdet/models/detectors/semi_base.py [hunk 4]
```diff
@@ -141,30 +154,21 @@ def loss_by_pseudo_instances(self,
...
-            for k, v in self.reweight_loss(
-                self.student.loss(batch_inputs, batch_data_samples),
-                unsup_weight).items()
-        }
-        return pseudo_loss
-
-    def filter_pseudo_instances(self,
-                                batch_data_samples: SampleList) -> SampleList:
-        """Filter invalid pseudo instances from teacher model."""
+    def filter_pseudo_instances_by_sizes(
```
Comment (ZwwWayne) [inline mmdet/models/detectors/semi_base.py:169]: by_sizes -> by size

File: mmdet/models/detectors/semi_base.py (hunk 1 above)
Comment (ZwwWayne) [reply mmdet/models/detectors/semi_base.py:72]: rename_loss is more common, it can be used in many places like DETR.

**Review by ZwwWayne — DISMISSED** [review]

File: mmdet/models/utils/misc.py [hunk 5]
```diff
@@ -535,3 +535,80 @@ def get_box_tensor(boxes: Union[Tensor, BaseBoxes]) -> Tensor:
...
+        losses (dict):  A dictionary of loss components.
+
+    Returns:
+            dict: A dictionary of loss components with prefix.
+    """
+    return {prefix + k: v for k, v in losses.items()}
+
+
+def reweight_loss(losses: dict, weight: float) -> dict:
+    """Reweight loss for different branches.
```
Comment (ZwwWayne) [inline mmdet/models/utils/misc.py:599]: Reweight losses in the dict by weight.
reweight_loss -> reweight_loss_dict

File: mmdet/models/utils/misc.py [hunk 6]
```diff
@@ -535,3 +535,80 @@ def get_box_tensor(boxes: Union[Tensor, BaseBoxes]) -> Tensor:
...
+        bboxes = data_samples.gt_instances.bboxes
+        if bboxes.shape[0] > 0:
+            w = bboxes[:, 2] - bboxes[:, 0]
+            h = bboxes[:, 3] - bboxes[:, 1]
+            data_samples.gt_instances = data_samples.gt_instances[
+                (w > min_bbox_wh[0]) & (h > min_bbox_wh[1])]
+    return batch_data_samples
+
+
+def rename_loss(prefix: str, losses: dict) -> dict:
```
Comment (ZwwWayne) [inline mmdet/models/utils/misc.py:585]: rename_loss(prefix: str, losses: dict) -> rename_loss_dict(prefix: str, losses: dict)

File: mmdet/models/utils/misc.py [hunk 7]
```diff
@@ -535,3 +535,80 @@ def get_box_tensor(boxes: Union[Tensor, BaseBoxes]) -> Tensor:
...
+    """Rename loss for different branches.
```
Comment (ZwwWayne) [inline mmdet/models/utils/misc.py:586]: Rename the key names in loss dict by adding a prefix.

File: mmdet/models/utils/misc.py [hunk 8]
```diff
@@ -535,3 +535,80 @@ def get_box_tensor(boxes: Union[Tensor, BaseBoxes]) -> Tensor:
     if isinstance(boxes, BaseBoxes):
         boxes = boxes.tensor
     return boxes
+
+
+def filter_instances_by_score(batch_data_samples: SampleList,
```
Comment (ZwwWayne) [inline mmdet/models/utils/misc.py:540]: filter_gt_instances_by_score

Comment (ZwwWayne) [reply mmdet/models/utils/misc.py:540]: we can unify filter_gt_instances_by_score and filter_gt_instances_by_size through
`filter_gt_instances(batch_data_samples, score_thr=None, wh_thr=None)`

File: mmdet/models/detectors/soft_teacher.py [hunk 9]
```diff
@@ -249,18 +241,18 @@ def rcnn_cls_loss_by_pseudo_instances(self, x: Tuple[Tensor],
                 rescale=False)
             bg_score = torch.cat(
                 [results.scores[:, -1] for results in results_list])
+            # cls_reg_targets[0] is labels
             neg_inds = cls_reg_targets[
                 0] == self.student.roi_head.bbox_head.num_classes
+            # cls_reg_targets[1] is label_weights
             cls_reg_targets[1][neg_inds] = bg_score[neg_inds].detach()
```
Comment (ZwwWayne) [inline mmdet/models/detectors/soft_teacher.py:248]: consider directly modifing the cls_reg_targets here to avoid lines 254&255

File: mmdet/models/utils/misc.py [hunk 10]
```diff
@@ -535,3 +535,105 @@ def get_box_tensor(boxes: Union[Tensor, BaseBoxes]) -> Tensor:
...
+            h = bboxes[:, 3] - bboxes[:, 1]
+            data_samples.gt_instances = data_samples.gt_instances[
+                (w > wh_thr[0]) & (h > wh_thr[1])]
+    return batch_data_samples
+
+
+def filter_gt_instances(batch_data_samples: SampleList,
+                        score_thr: float = None,
+                        wh_thr: tuple = None):
+    """Filter gt-instances by score and size.
```
Comment (ZwwWayne) [inline mmdet/models/utils/misc.py:588]: Filter ground truth (GT) instances by score and/or size

File: mmdet/models/utils/misc.py [hunk 11]
```diff
@@ -535,3 +535,105 @@ def get_box_tensor(boxes: Union[Tensor, BaseBoxes]) -> Tensor:
...
+            `gt_instance`, `gt_panoptic_seg` and `gt_sem_seg`.
+        score_thr (float): The score filter threshold.
+        wh_thr (tuple):  Minimum width and height of bbox.
+
+    Returns:
+        SampleList: The Data Samples filtered by score.
+    """
+
+    if score_thr is not None:
+        batch_data_samples = filter_gt_instances_by_score(
```
Comment (ZwwWayne) [inline mmdet/models/utils/misc.py:602]: we do not need to keep these two functions anymore so that we can simplify the code and logic

**Review by ZwwWayne — APPROVED** [review]
